### ISSUE TITLE
3.9.2 master: hourly fork / 30 seconds down time (all clients)

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#323]
At the begginning of every hour (with few seconds precision) all clients timeout for ~30 seconds as follows:

~~~~
Oct 31 20:00:08 debmain mfsmount[9250]: master: tcp recv error: ETIMEDOUT (Operation timed out)
Oct 31 20:00:09 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:10 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:11 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:12 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:13 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:14 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:15 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:16 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:17 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:18 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:20 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:21 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:22 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:23 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:24 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:25 debmain mfsmount[9250]: master: register error (read header: ETIMEDOUT (Operation timed out))
Oct 31 20:00:26 debmain mfsmount[9250]: registered to master (session id #2437)
~~~~

Inspection of master log revealed:

~~~~
Oct 31 20:00:00 debstor mfsmaster[20991]: fork failed: ENOMEM (Cannot allocate memory)
Oct 31 20:00:00 debstor mfsmaster[20991]: mfsmaster[20991]: fork failed: ENOMEM (Cannot allocate memory)
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.2) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.76) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.2) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.76) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.204) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.7) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.75) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.2) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.76) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.204) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.7) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.75) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.2) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.76) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.204) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.7) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.75) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.2) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.76) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.204) has been closed by peer
Oct 31 20:00:26 debstor mfsmaster[20991]: connection with client(ip:192.168.0.7) has been closed by peer
[...]
~~~~

Fork fails because master uses about 50% of server's RAM and only about 40% is available. Master is started with `LOCK_MEMORY = 1`.

LizardFS compiled with _libJudy_ but without _tcmalloc_ (due to performance reasons).

I do not recall such problem on 2.6.0. It would be nice to avoid hourly forking as fork needs twice as much RAM as already allocated by master.

**Comment by psarna** [comment]
This is not 3.9.2 issue, previous versions of LizardFS and actually any executable file in the world could have the same issue with fork.
The problem is that Linux assumes, that the forked process will need as much memory as the original one, which is usually not true because of copy-on-write.
Issue can be solved with setting overcommit_memory option in the kernel,
which will let processes fork even if they *could* potentially need more memory than amount available in the system:
~~~
echo 1 > /proc/sys/vm/overcommit_memory
~~~

**Comment by Zorlin** [comment]
Alternately, you could also consider RAM requirements as being twice the
size of the metadata set. At my previous employer that was the answer -
gobs of RAM with plenty of room for expansion.

**Comment by onlyjob** [comment]
I'm surprised that `vm.overcommit_memory=1` is _required_ in such situation. I'm not sure if I'm comfortable with that. Besides default is `vm.overcommit_memory=0`

By why hourly fork in first place?? Is it really necessary? What for?

@Zorlin, amount of RAM is roughly twice the size of metadata. Co-located chunkserver takes another 10% of RAM. 60% RAM utilisation is not bad and it is certainly not nice to require twice as much RAM so master could fork.
I expect process to be able to allocate all the memory it needs.

I consider this issue to be a bug in master's memory management hence I respectfully request to reopen.

**Comment by psarna** [comment]
Forks are used for hourly metadata dumps. The design comes frm MFS1.6 and is definitely not flawless, but rewriting the mechanism without forks is not considered a priority right now.

**Comment by onlyjob** [comment]
Thanks for reopening. I agree that it is indeed a low priority issue (especially is `vm.overcommit_memory=1` helps) but it would be nice to have an agenda to get rid of hourly forks in (long-term) TODO list.

Meanwhile here is an article explaining some pitfalls of overcommitting: http://www.etalabs.net/overcommit.html

Thanks.

**Comment by biocyberman** [comment]
@onlyjob Could you explain a bit more of how you overcome this problem for now?

I am running 3.10 and writing from one client to one of chunkservers is timed out. Shutting down processes and services to free up memory on the mfsmaster seems to help. But what is a better solution other than the obvious: buy more RAM?

**Comment by onlyjob** [comment]
On 3.9.4 master used about 51% of RAM on my server (where buying more RAM is not an option).
For some time I've used `vm.overcommit_memory=1` (see `sysctl(8)`) with some extra swap space.

3.10.1 optimised RAM usage so master uses a little less than 50% of RAM and does not exhibit the problem.

**Comment by nickcoons** [comment]
I have an instance of 3.13RC1, and I encounter this issue 1-2 times daily as described above where it happens a few seconds after the hour (though not the same hour each day). The master server has 128GB of RAM and only 5GB being used for the master service managing approximately 14m chunks. I have `vm.overcommit_memory=1` set as described here, and doing that seems to have reduced the occurrence (it used to happen once every few hours).

**Comment by AmokHuginnsson** [comment]
If you have slow disk where your metadata directory is located then it can also exaggerate this behavior. It could be alleviated completely but that would require rewrite in master that would allow storing changelogs and metadata dumps on separated devices.

**Comment by nickcoons** [comment]
For my setup, metadata is stored on a pair of SSDs in a RAID 1 array. I don't have spinning disks anywhere in any system in this cluster.

Is there no way, as the system is built now, to configure it such that the metadata master service is not blocked while this hourly writing takes place? As I understood from reading above, the process that does this hourly is a fork. Does the forked process not run in parallel such that the parent service is blocked until the forked process completes?

Alternatively, are the hourly metadata dumps hard-coded, or are they configurable so that I could cause them to happen at specific times? If there has to be a regular outage because of the way this works, I'd much prefer to be able to schedule that possible outage rather than it happening randomly during business hours.

**Comment by AmokHuginnsson** [comment]
Yes. Metadata dump periods are hardcoded now.
I am guessing that your clients are blocked because changelog appending operation that happens on every metadata update is waiting on IO because metadata dump is being written at the moment.
Maybe it could be possible to reconfigure "something" on system level so that concurrent writes have more "even" access to the drive. But like I said I am just guessing here. Proper fix (that could possibly involve changelog/metadata dump drive separation I mentioned in my previous post) will happen after 3.13 actual release.
